scripts/mbedtls_dev/bignum_mod_raw.py

```diff
@@ -107,6 +108,70 @@ def result(self) -> List[str]:
 
 # BEGIN MERGE SLOT 6
 
+class BignumModRawConvertRep(bignum_common.ModOperationCommon,
+                             BignumModRawTarget):
+    # This is an abstract class, it's ok to have unimplemented methods.
+    #pylint: disable=abstract-method
+    """Test cases for representation conversion."""
+    arity = 1
+
+    def __init__(self, val_n: str, val_a: str, bits_in_limb: Optional[int],
+                 rep: bignum_common.ModulusRepresentation) -> None:
+        if bits_in_limb is None:
+            super().__init__(val_n=val_n, val_a=val_a)
+        else:
+            self.input_style = "arch_split"
+            super().__init__(val_n=val_n, val_a=val_a, bits_in_limb=bits_in_limb)
+        self.rep = rep
+
+    def arguments(self) -> List[str]:
+        return ([bignum_common.quote_str(self.arg_n), self.rep.symbol(),
+                 bignum_common.quote_str(self.arg_a)] +
+                self.result())
+
+    def description(self) -> str:
+        base = super().description()
+        mod_with_rep = 'mod({})'.format(self.rep.name)
+        return base.replace('mod', mod_with_rep, 1)
+
+    @classmethod
+    def generate_function_tests(cls) -> Iterator[test_case.TestCase]:
+        representations = \
+            bignum_common.ModulusRepresentation.supported_representations()
+        for rep in representations:
+            if rep is bignum_common.ModulusRepresentation.MONTGOMERY:
+                limb_sizes = cls.limb_sizes #type: Union[List[int], List[None]]
+            else:
+                limb_sizes = [None] # no dependency on limb size
+            for n in cls.moduli:
+                for a in cls.input_values:
+                    for bil in limb_sizes:
+                        test_object = cls(n, a, bil, rep)
+                        if test_object.is_valid:
+                            yield test_object.create_test_case()
+
+class BignumModRawCanonicalToModulusRep(BignumModRawConvertRep):
+    """Test cases for mpi_mod_raw_canonical_to_modulus_rep."""
+    test_function = "mpi_mod_raw_canonical_to_modulus_rep"
+    test_name = "Rep canon->mod"
+
+    def result(self) -> List[str]:
+        result = self.convert_from_canonical(self.int_a, self.rep)
+        return [self.format_result(result)]
+
+class BignumModRawModulusToCanonicalRep(BignumModRawConvertRep):
+    """Test cases for mpi_mod_raw_modulus_to_canonical_rep."""
+    test_function = "mpi_mod_raw_modulus_to_canonical_rep"
+    test_name = "Rep mod->canon"
```

Comment:
The argument in favour of Python was that it would make reviewing test case generation simpler. Well, that simply isn't the case. These two classes are trivial, but their superclass `BignumModRawConvertRep` is much more complicated - it took me ages to convince myself that we would get the right output with an `arch_split` for Montgomery conversion, and I had to checkout the code, run it, and examine the generated files to be certain. Yes, I (generally) always do that anyway, but this is so far from "making reviewer's life simpler" as I can imagine.

Anyway, it's in test code, so it's fine.

Reply:
I agree. I think we need to spend some time after the holidays digesting the lessons learned about bignum test generation and doing some redesign.